# F21NL (2025-26) – Coursework 1
Mercy Nthwia-H00292623 &
Christa Mary Veluthedath Jacob- H00479451

In [1]:
!wget -O wikitext-filtered-full.zip "https://www.dropbox.com/scl/fi/ibd4cmixckghx6hhb361c/wikitext-filtered-full.zip?rlkey=q71cebf0k5fvvwhmcntoswzhq&dl=1"
!wget -O wikitext-filtered-10k.zip "https://www.dropbox.com/scl/fi/ek174r3sg7qjx0aa9atop/wikitext-filtered-10k.zip?rlkey=zy6jqxv6qsc16lr9qm3ki9uhf&dl=1"

--2025-10-20 11:34:15--  https://www.dropbox.com/scl/fi/ibd4cmixckghx6hhb361c/wikitext-filtered-full.zip?rlkey=q71cebf0k5fvvwhmcntoswzhq&dl=1
Resolving www.dropbox.com (www.dropbox.com)... 162.125.1.18, 2620:100:6016:18::a27d:112
Connecting to www.dropbox.com (www.dropbox.com)|162.125.1.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://uc316f041f8b4736a298729cca82.dl-eu.dropboxusercontent.com/cd/0/inline/CznxwACkd1yJ8DAdBO90iVthhSISzwgPIMUji3S5ceGU01hvrkNETARPCrGgDi2bI8nQQHTFSe3nrMCJT5WU1KmL73yHnn55RMp2scdzgECkvRm-tE4O9svymak9gYO1AD-OHBm8b4qNOjtPrQDiI5z2/file?dl=1# [following]
--2025-10-20 11:34:16--  https://uc316f041f8b4736a298729cca82.dl-eu.dropboxusercontent.com/cd/0/inline/CznxwACkd1yJ8DAdBO90iVthhSISzwgPIMUji3S5ceGU01hvrkNETARPCrGgDi2bI8nQQHTFSe3nrMCJT5WU1KmL73yHnn55RMp2scdzgECkvRm-tE4O9svymak9gYO1AD-OHBm8b4qNOjtPrQDiI5z2/file?dl=1
Resolving uc316f041f8b4736a298729cca82.dl-eu.dropboxusercontent.com (uc316f041f8b4736a298729cca82.dl-eu.dropbo

In [2]:
!unzip wikitext-filtered-full.zip
!unzip wikitext-filtered-10k.zip

Archive:  wikitext-filtered-full.zip
   creating: wikitext-filtered-full/
  inflating: wikitext-filtered-full/dataset_info.json  
  inflating: wikitext-filtered-full/state.json  
  inflating: wikitext-filtered-full/data-00000-of-00001.arrow  
Archive:  wikitext-filtered-10k.zip
   creating: wikitext-filtered-10k/
  inflating: wikitext-filtered-10k/dataset_info.json  
  inflating: wikitext-filtered-10k/state.json  
  inflating: wikitext-filtered-10k/data-00000-of-00001.arrow  


In [3]:
!pip install datasets
import datasets

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.1/75.1 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 506.3/506.3 kB 19.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.7/119.7 kB 12.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 146.7/146.7 kB 14.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.9/193.9 kB 19.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 69.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 242.4/242.4 kB 24.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 221.6/221.6 kB 22.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 377.3/377.3 kB 33.6 MB/s eta 0:00:00


In [4]:
## step 1
# Reference: Official Hugging Face Datasets library
# https://huggingface.co/docs/datasets
from datasets import load_dataset, Dataset

def load_dataset():
  wikitext_small = "wikitext-filtered-10k"
  wikitext_large = "wikitext-filtered-full"

  dataset_small = Dataset.load_from_disk(wikitext_small)
  dataset_large = Dataset.load_from_disk(wikitext_large)
  print("wikitext_small: {} docs, wikitext_large: {} docs".format(len(dataset_small), len(dataset_large)))
  return dataset_small, dataset_large

wikitext_small, wikitext_large = load_dataset()

wikitext_small: 10000 docs, wikitext_large: 859955 docs


In [5]:
!pip install gensim

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.9/27.9 MB 84.2 MB/s eta 0:00:00


In [6]:
# DATA EXPLORATION BEFORE PREPROCESSING
# 1. Dataset Size
num_docs_small = len(wikitext_small)
num_docs_large = len(wikitext_large)

print("DATASET SIZE:")
print(f"  Documents (small): {num_docs_small:,}")
print(f"  Documents (large): {num_docs_large:,}")
print()

# 2. Raw token counts
total_tokens_small = sum(len(doc.split()) for doc in wikitext_small['text'])
total_tokens_large = sum(len(doc.split()) for doc in wikitext_large['text'])

print("TOKEN COUNTS:")
print(f"  Total tokens (small): {total_tokens_small:,}")
print(f"  Total tokens (large): {total_tokens_large:,}")
print(f"  Avg doc length (small): {total_tokens_small/num_docs_small:.1f} tokens")
print(f"  Avg doc length (large): {total_tokens_large/num_docs_large:.1f} tokens")
print()

# 3. Raw vocabulary size
raw_vocab_small = set(word.lower() for doc in wikitext_small['text'] for word in doc.split())
raw_vocab_large = set(word.lower() for doc in wikitext_large['text'] for word in doc.split())

print("RAW VOCABULARY SIZE:")
print(f"  Vocabulary (small): {len(raw_vocab_small):,}")
print(f"  Vocabulary (large): {len(raw_vocab_large):,}")
print()

# 4. Sample documents
print("SAMPLE DOCUMENTS:")
print(f"  Small corpus [0]: {wikitext_small['text'][0]}")
print(f"  Large corpus [0]: {wikitext_large['text'][0]}")
print()

# 5. Data quality checks
empty_small = sum(1 for doc in wikitext_small['text'] if len(doc.strip()) == 0)
empty_large = sum(1 for doc in wikitext_large['text'] if len(doc.strip()) == 0)

print("DATA QUALITY:")
print(f"  Empty documents (small): {empty_small}")
print(f"  Empty documents (large): {empty_large}")
print()

# 6. Stopword analysis
from gensim.parsing.preprocessing import STOPWORDS

# Number of stopwords in vocabularies
stopwords_in_vocab_small = [w for w in raw_vocab_small if w in STOPWORDS]
stopwords_in_vocab_large = [w for w in raw_vocab_large if w in STOPWORDS]

print("STOPWORDS IN VOCABULARY:")
print(f"  Stopwords (small): {len(stopwords_in_vocab_small)} out of {len(STOPWORDS)}")
print(f"  First 10: {stopwords_in_vocab_small[:10]}")
print()
print(f"  Stopwords (large): {len(stopwords_in_vocab_large)} out of {len(STOPWORDS)}")
print(f"  First 10: {stopwords_in_vocab_large[:10]}")
print()

# Missing stopwords
all_stopwords = set(STOPWORDS)

stopwords_small_set = set(stopwords_in_vocab_small)
stopwords_large_set = set(stopwords_in_vocab_large)

missing_small = all_stopwords - stopwords_small_set
missing_large = all_stopwords - stopwords_large_set

print(f"Missing from small corpus: {sorted(missing_small)}")
print(f"Missing from large corpus: {sorted(missing_large)}")

DATASET SIZE:
  Documents (small): 10,000
  Documents (large): 859,955

TOKEN COUNTS:
  Total tokens (small): 1,150,954
  Total tokens (large): 99,185,686
  Avg doc length (small): 115.1 tokens
  Avg doc length (large): 115.3 tokens

RAW VOCABULARY SIZE:
  Vocabulary (small): 48,414
  Vocabulary (large): 546,768

SAMPLE DOCUMENTS:
  Small corpus [0]: Senjō no Valkyria 3 : Unrecorded Chronicles ( Japanese : 戦場のヴァルキュリア3 , lit . Valkyria of the Battlefield 3 ) , commonly referred to as Valkyria Chronicles III outside Japan , is a tactical role @-@ playing video game developed by Sega and Media.Vision for the PlayStation Portable . Released in January 2011 in Japan , it is the third game in the Valkyria series . Employing the same fusion of tactical and real @-@ time gameplay as its predecessors , the story runs parallel to the first game and follows the " Nameless " , a penal military unit serving the nation of Gallia during the Second Europan War who perform secret black operations and a

In [7]:
# Step 2
# References:
# Mikolov, T., Chen, K., Corrado, G., & Dean, J. (2013).
#   Efficient Estimation of Word Representations in Vector Space.
#   arXiv preprint arXiv:1301.3781.
#  https://arxiv.org/abs/1301.3781

# Gensim Documentation – Word2Vec and Preprocessing Utilities:
#   https://radimrehurek.com/gensim/models/word2vec.html
#   https://radimrehurek.com/gensim/utils.html
#
# NLTK Documentation – WordNet Lemmatizer:
#   https://www.nltk.org/api/nltk.stem.wordnet.html

# The preprocessing uses:
#   - simple_preprocess (Gensim)
#   - remove_stopwords (Gensim)
#   - WordNetLemmatizer (NLTK)
#   for text cleaning and normalization before training Word2Vec models.

from gensim.utils import simple_preprocess
from gensim.parsing.preprocessing import remove_stopwords
from nltk.stem import WordNetLemmatizer
from gensim.models import Word2Vec
import nltk

# Required NLTK data download
nltk.download('wordnet')
nltk.download('omw-1.4')

# ---------------------------------------------------------
# Preprocessing function with lemmatization
# ---------------------------------------------------------
def preprocess_texts(dataset, use_lemmatization=True):
    """
    Cleans and tokenizes the dataset using Gensim
    - Removes stopwords
    - Tokenizes words with simple_preprocess
    - Lemmatizes using NLTK
    Returns: list of token lists
    """
    lemmatizer = WordNetLemmatizer() if use_lemmatization else None
    processed = []

    for text in dataset['text']:
        if isinstance(text, str) and text.strip():
            # Remove stopwords
            cleaned = remove_stopwords(text.lower())
            # Tokenize using gensim's simple_preprocess
            tokens = simple_preprocess(cleaned, deacc=True)

            # Apply lemmatization if enabled
            if use_lemmatization:
                tokens = [lemmatizer.lemmatize(token) for token in tokens]

            processed.append(tokens)

    return processed

# ---------------------------------------------------------
# Preprocess both corpora
# ---------------------------------------------------------
train_small = preprocess_texts(wikitext_small, use_lemmatization=True)
train_large = preprocess_texts(wikitext_large, use_lemmatization=True)

print(f"Sample tokens (small): {train_small[0][:10]}")
print(f"Sample tokens (large): {train_large[0][:10]}")

# ---------------------------------------------------------
# Define baseline parameters
# ---------------------------------------------------------
params = dict(vector_size=50, window=5, min_count=5, epochs=5)

# ---------------------------------------------------------
# Train models
# ---------------------------------------------------------
print("Training Word2Vec on wikitext_small...")
model_small = Word2Vec(sentences=train_small, **params)
print("Small model trained.")

print("Training Word2Vec on wikitext_large...")
model_large = Word2Vec(sentences=train_large, **params)
print("Large model trained.")

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


Sample tokens (small): ['senjo', 'valkyria', 'unrecorded', 'chronicle', 'japanese', '戦場のウァルキュリア', 'lit', 'valkyria', 'battlefield', 'commonly']
Sample tokens (large): ['senjo', 'valkyria', 'unrecorded', 'chronicle', 'japanese', '戦場のウァルキュリア', 'lit', 'valkyria', 'battlefield', 'commonly']
Training Word2Vec on wikitext_small...
Small model trained.
Training Word2Vec on wikitext_large...
Large model trained.


In [8]:
# ---------------------------------------------------------
# Model verification
# ---------------------------------------------------------
# References:
# Mikolov, T., Chen, K., Corrado, G., & Dean, J. (2013).
#   Efficient Estimation of Word Representations in Vector Space.
#   arXiv preprint arXiv:1301.3781.
#   https://arxiv.org/abs/1301.3781

# Řehůřek, R., & Sojka, P. (2010).
#   Software Framework for Topic Modelling with Large Corpora.
#   In Proceedings of the LREC 2010 Workshop on New Challenges for NLP Frameworks.
#   ELRA, Valletta, Malta.
#   https://radimrehurek.com/gensim/

# Gensim Documentation – Word2Vec Similarity and Nearest Neighbours:
#   https://radimrehurek.com/gensim/models/keyedvectors.html

# The verification checks confirm:
#   - Vocabulary size and embedding dimension integrity
#   - Sanity-check cosine similarities between related words
#   - Nearest-neighbour relationships for semantic consistency

print("MODEL VERIFICATION")

# 1. Vocabulary sizes
print(f"\nVocabulary size (small): {len(model_small.wv)}")
print(f"Vocabulary size (large): {len(model_large.wv)}")

# 2. Token counts
total_tokens_small = sum(len(doc) for doc in train_small)
total_tokens_large = sum(len(doc) for doc in train_large)

print(f"\nToken count (small): {total_tokens_small:,}")
print(f"Token count (large): {total_tokens_large:,}")

# 3. Vector dimensions
print(f"\nVector dimensions (small): {model_small.wv.vector_size}")
print(f"Vector dimensions (large): {model_large.wv.vector_size}")

# 4. Test basic similarity (sanity check)
try:
    sim_small = model_small.wv.similarity("driver", "car")
    print(f"\nSimilarity 'driver' vs 'car' (small): {sim_small:.4f}")
except KeyError:
    print("\nWord 'driver' or 'car' not in vocabulary (small)")

try:
    sim_large = model_large.wv.similarity("driver", "car")
    print(f"Similarity 'driver' vs 'car' (large): {sim_large:.4f}")
except KeyError:
    print("Word 'driver' or 'car' not in vocabulary (large)")
print()

# 5. Neighbours
#checks:
print("\nNeighbors of 'driver' (small):")
print(model_small.wv.most_similar("driver", topn=5))

print("\nNeighbors of 'driver' (large):")
print(model_large.wv.most_similar("driver", topn=5))

print("\nNeighbors of 'mother' (small):")
print(model_small.wv.most_similar("mother", topn=5))

print("\nNeighbors of 'mother' (large):")
print(model_large.wv.most_similar("mother", topn=5))

MODEL VERIFICATION

Vocabulary size (small): 12291
Vocabulary size (large): 148659

Token count (small): 505,584
Token count (large): 43,573,514

Vector dimensions (small): 50
Vector dimensions (large): 50

Similarity 'driver' vs 'car' (small): 0.9833
Similarity 'driver' vs 'car' (large): 0.8010


Neighbors of 'driver' (small):
[('finally', 0.9962689280509949), ('recovered', 0.9958595037460327), ('organised', 0.9956149458885193), ('glider', 0.9955239295959473), ('parade', 0.9954498410224915)]

Neighbors of 'driver' (large):
[('car', 0.8010475635528564), ('mclaren', 0.789621889591217), ('benetton', 0.7810556292533875), ('minardi', 0.7688940167427063), ('ferrari', 0.7670093178749084)]

Neighbors of 'mother' (small):
[('father', 0.9903742671012878), ('parent', 0.9804277420043945), ('yue', 0.9725046157836914), ('woman', 0.9697748422622681), ('zhou', 0.966592013835907)]

Neighbors of 'mother' (large):
[('grandmother', 0.8914514780044556), ('aunt', 0.8797212243080139), ('wife', 0.86048674583

Both models were trained successfully on their respective corpora and produce valid 50-dimensional word embeddings.
The larger model, trained on a more extensive dataset, demonstrates semantically coherent relationships (e.g., driver–car, mother–daughter).
In contrast, the smaller model shows inflated similarity scores and noisier neighbourhoods, indicating limited contextual diversity and weaker generalisation.

In [9]:
model_small.save("word2vec_small.model")
model_large.save("word2vec_large.model")


In [10]:
# Step 3: Semantic Relatedness Task
# Reference:
# References:
# Agirre, E., Alfonseca, E., Hall, K., Kravalova, J., Paşca, M., & Soroa, A. (2009).
#   A Study on Similarity and Relatedness Using Distributional and WordNet-based Approaches.
#   In Proceedings of NAACL-HLT 2009, Boulder, Colorado.
#   https://gabrilovich.com/resources/data/wordsim353/wordsim353.zip
#
# Mikolov, T., Chen, K., Corrado, G., & Dean, J. (2013).
#   Efficient Estimation of Word Representations in Vector Space.
#   arXiv preprint arXiv:1301.3781.
#
# Gensim Documentation – Word2Vec and Similarity Computations:
#   https://radimrehurek.com/gensim/models/keyedvectors.html
#
# SciPy Documentation – Cosine Distance Function:
#   https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cosine.html
# ------------------------------------------------------------

import requests, zipfile, io, pandas as pd
from scipy.spatial.distance import cosine

# ------------------------------------------------------------
# 1. Download and extract the WordSim-353 dataset
# ------------------------------------------------------------
url = "https://gabrilovich.com/resources/data/wordsim353/wordsim353.zip"
response = requests.get(url)
z = zipfile.ZipFile(io.BytesIO(response.content))
z.extractall("wordsim353")
print(" WordSim-353 downloaded and extracted successfully.")

# ------------------------------------------------------------
# 2. Load the 'combined.csv' file into a DataFrame
# ------------------------------------------------------------
ws353 = pd.read_csv("wordsim353/combined.csv")

# ------------------------------------------------------------
# 3. Define cosine similarity function (Gensim-compatible)
# ------------------------------------------------------------
def cosine_similarity(word1, word2, model):
    """
    Compute cosine similarity between two words.
    Works for both Word2Vec (trained) and KeyedVectors (pretrained) models.
    """
    # Determine key set depending on model type
    if hasattr(model, 'wv'):
        key_index = model.wv.key_to_index
        get_vec = model.wv.__getitem__
    else:
        key_index = model.key_to_index
        get_vec = model.__getitem__

    if word1 in key_index and word2 in key_index:
        return 1 - cosine(get_vec(word1), get_vec(word2))
    else:
        return None

# ------------------------------------------------------------
# 4. Compute similarities for required pairs
# ------------------------------------------------------------
pairs = [
    ("plane", "car"),
    ("planet", "sun"),
    ("cup", "article"),
    ("sugar", "approach")
]

print("\n=== Cosine Similarities (wikitext_small) ===")
for w1, w2 in pairs:
    sim = cosine_similarity(w1, w2, model_small)
    print(f"{w1:>10s} / {w2:<10s} : {sim}")

print("\n=== Cosine Similarities (wikitext_large) ===")
for w1, w2 in pairs:
    sim = cosine_similarity(w1, w2, model_large)
    print(f"{w1:>10s} / {w2:<10s} : {sim}")


 WordSim-353 downloaded and extracted successfully.

=== Cosine Similarities (wikitext_small) ===
     plane / car        : 0.9660117030143738
    planet / sun        : 0.9957811832427979
       cup / article    : 0.6256794333457947
     sugar / approach   : 0.9878817200660706

=== Cosine Similarities (wikitext_large) ===
     plane / car        : 0.5781422853469849
    planet / sun        : 0.5768535733222961
       cup / article    : 0.02065277099609375
     sugar / approach   : -0.21420443058013916


| **Word Pair**        | **wikitext_small** | **wikitext_large** | **Interpretation**                                                                                                                                       |
| -------------------- | ------------------ | ------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **plane / car**      | 0.959              | 0.557              | Both are vehicles, but not synonyms. The smaller corpus overestimates similarity, while the large corpus captures more realistic semantic distance.      |
| **planet / sun**     | 0.994              | 0.589              | Both are celestial objects. The large model correctly shows moderate similarity, while the small model exaggerates due to poor contextual understanding. |
| **cup / article**    | 0.610              | 0.020              | The small model falsely links unrelated words; the large model correctly indicates near-zero similarity.                                                 |
| **sugar / approach** | 0.976              | -0.214             | The small model confuses unrelated terms, producing a spurious high similarity; the large model correctly identifies them as dissimilar.                 |


In [11]:
#checks:
# ============================================================
# GET NEIGHBORS FOR EXPLANATION SUPPORT
# ============================================================

pairs = [
    ("plane", "car"),
    ("planet", "sun"),
    ("cup", "article"),
    ("sugar", "approach")
]

print("WIKITEXT_SMALL - Top 5 Neighbors:\n")
for w1, w2 in pairs:
    print(f"{w1} / {w2}:")
    if w1 in model_small.wv:
        print(f"  '{w1}': {[n[0] for n in model_small.wv.most_similar(w1, topn=5)]}")
    if w2 in model_small.wv:
        print(f"  '{w2}': {[n[0] for n in model_small.wv.most_similar(w2, topn=5)]}")
    print()

print("\nWIKITEXT_LARGE - Top 5 Neighbors:\n")
for w1, w2 in pairs:
    print(f"{w1} / {w2}:")
    if w1 in model_large.wv:
        print(f"  '{w1}': {[n[0] for n in model_large.wv.most_similar(w1, topn=5)]}")
    if w2 in model_large.wv:
        print(f"  '{w2}': {[n[0] for n in model_large.wv.most_similar(w2, topn=5)]}")
    print()

WIKITEXT_SMALL - Top 5 Neighbors:

plane / car:
  'plane': ['disaster', 'rescue', 'threatened', 'parliamentary', 'arrested']
  'car': ['suggestion', 'reputation', 'donald', 'quest', 'fast']

planet / sun:
  'planet': ['moon', 'skin', 'creature', 'guilty', 'gill']
  'sun': ['moon', 'difficult', 'writes', 'difference', 'planet']

cup / article:
  'cup': ['championship', 'rugby', 'fa', 'victory', 'won']
  'article': ['doubt', 'revealed', 'discussed', 'earnest', 'crab']

sugar / approach:
  'sugar': ['tribbles', 'plastic', 'motion', 'addition', 'tend']
  'approach': ['ring', 'opposite', 'display', 'sitting', 'coat']


WIKITEXT_LARGE - Top 5 Neighbors:

plane / car:
  'plane': ['airplane', 'helicopter', 'fireflash', 'flight', 'wreckage']
  'car': ['motorcycle', 'truck', 'suv', 'driver', 'cab']

planet / sun:
  'planet': ['galaxy', 'pluto', 'earth', 'planetary', 'jupiter']
  'sun': ['equinox', 'jupiter', 'moon', 'bright', 'luminous']

cup / article:
  'cup': ['tournament', 'qualifier', 'linf

In [12]:
# ============================================================
# STEP 4 – Evaluate Semantic Relatedness (Full 353 pairs)
# ============================================================
# References:
# Agirre et al., 2009 – WordSim-353 Dataset
# https://gabrilovich.com/resources/data/wordsim353/wordsim353.zip
#
# Mikolov et al., 2013 – Efficient Estimation of Word Representations in Vector Space
# https://arxiv.org/abs/1301.3781
#
# SciPy Documentation – Spearman Rank-Order Correlation
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.spearmanr.html
#
# Gensim Documentation – KeyedVectors & similarity computations
# https://radimrehurek.com/gensim/models/keyedvectors.html
# ============================================================

from scipy.spatial.distance import cosine
from scipy.stats import spearmanr
import pandas as pd
import numpy as np

# ------------------------------------------------------------
# 1. Compute cosine similarities for *all* 353 word pairs
# ------------------------------------------------------------
def compute_cosine_similarities(model, dataset):
    """
    Computes cosine similarity for every WordSim-353 pair.
    Returns a new DataFrame with a 'cosine_similarity' column.
    Out-of-vocabulary (OOV) pairs get similarity = 0
    to ensure 353 values per model.
    """
    similarities = []
    # choose correct vector accessor
    if hasattr(model, "wv"):
        key_index = model.wv.key_to_index
        get_vec = model.wv.__getitem__
    else:
        key_index = model.key_to_index
        get_vec = model.__getitem__

    for _, row in dataset.iterrows():
        w1, w2 = row["Word 1"].lower(), row["Word 2"].lower()
        if w1 in key_index and w2 in key_index:
            sim = 1 - cosine(get_vec(w1), get_vec(w2))
        else:
            sim = 0.0        # ensure 353 numbers
        similarities.append(sim)

    dataset = dataset.copy()
    dataset["cosine_similarity"] = similarities
    return dataset

# ------------------------------------------------------------
# 2. Evaluate Spearman correlation (full 353 pairs)
# ------------------------------------------------------------
def evaluate_spearman_full(model, dataset, label="Model"):
    benchmark_df = compute_cosine_similarities(model, dataset)

    # drop NaNs if any numerical issues arise
    valid_pairs = benchmark_df.dropna(subset=["cosine_similarity"])

    spearman_corr, spearman_pval = spearmanr(
        valid_pairs["Human (mean)"],
        valid_pairs["cosine_similarity"]
    )

    print(f"=== {label}: Spearman rank-order correlation ===")
    print(f"ρ = {spearman_corr:.4f}")
    print(f"p-value = {spearman_pval:.4e}")
    print(f"Valid pairs used: {len(valid_pairs)} / {len(dataset)}\n")

    return benchmark_df, spearman_corr, spearman_pval

# ------------------------------------------------------------
# 3. Run evaluation for both models
# ------------------------------------------------------------
benchmark_small, rho_small, p_small = evaluate_spearman_full(model_small, ws353, "wikitext_small")
benchmark_large, rho_large, p_large = evaluate_spearman_full(model_large, ws353, "wikitext_large")

# ------------------------------------------------------------
# 4. Show example subset of key pairs
# ------------------------------------------------------------
word_pairs = [
    ('plane', 'car'),
    ('planet', 'sun'),
    ('cup', 'article'),
    ('sugar', 'approach')
]

filtered_df = ws353[
    ws353.apply(
        lambda row: (row["Word 1"], row["Word 2"]) in word_pairs
                    or (row["Word 2"], row["Word 1"]) in word_pairs,
        axis=1
    )
]

print("Example Word Pair Results (from large model):")
print("-" * 50)
for _, row in filtered_df.iterrows():
    w1, w2 = row["Word 1"], row["Word 2"]
    human = row["Human (mean)"]
    cosine_val = compute_cosine_similarities(model_large, pd.DataFrame([row]))["cosine_similarity"].iloc[0]
    print(f"{w1:>10s} – {w2:<10s} | Human: {human:>5.2f} | Cosine: {cosine_val:>6.3f}")


=== wikitext_small: Spearman rank-order correlation ===
ρ = -0.0011
p-value = 9.8316e-01
Valid pairs used: 353 / 353

=== wikitext_large: Spearman rank-order correlation ===
ρ = 0.6194
p-value = 8.7948e-39
Valid pairs used: 353 / 353

Example Word Pair Results (from large model):
--------------------------------------------------
     plane – car        | Human:  5.77 | Cosine:  0.578
    planet – sun        | Human:  8.02 | Cosine:  0.577
       cup – article    | Human:  2.40 | Cosine:  0.021
     sugar – approach   | Human:  0.88 | Cosine: -0.214


| Model              | Spearman ρ | p-value      | Valid Pairs | Interpretation                                                                                       |
| ------------------ | ---------- | ------------ | ----------- | ---------------------------------------------------------------------------------------------------- |
| **wikitext_small** | −0.0027    | 9.60 × 10⁻¹  | 353 / 353   | No meaningful correlation; model similarities are random relative to human scores                    |
| **wikitext_large** | **0.6138** | 6.27 × 10⁻³⁸ | 353 / 353   | Strong, statistically significant positive correlation; good semantic alignment with human intuition |


b. What is the value of the Spearman correlation coefficients computed in Step 4?
wikitext_small : ρ = −0.0027 (p ≈ 0.96, 353 pairs)

wikitext_large : ρ = 0.6138 (p < 0.001, 353 pairs)


c. How do you interpret each coefficient value with respect to the word-similarity task? Are the coefficient values for the two vector-space models different, and if so, why?

The Spearman rank-order correlation measures how well the model’s predicted word-similarity rankings agree with human judgments from the WordSim-353 benchmark.

For wikitext_small, ρ ≈ 0 shows no meaningful correlation. The model’s similarity predictions are essentially random, due to the small corpus’s limited context and sparse vocabulary coverage.

For wikitext_large, ρ ≈ 0.61 indicates a strong, statistically significant positive correlation. This means the large model’s semantic relationships (for example, plane–car or planet–sun) are closely aligned with human intuition.

The difference arises because the larger corpus exposes the model to far more co-occurrence patterns and richer linguistic contexts, allowing it to learn more accurate and generalisable word representations.

In [13]:
# ============================================================
# STEP 5: Evaluate Pretrained Word2Vec (Google News)
# ============================================================
# Reference:
# Mikolov et al., 2013 – Efficient Estimation of Word Representations in Vector Space
# Gensim Downloader Documentation:
#   https://radimrehurek.com/gensim/downloader.html
# ============================================================

import gensim.downloader as api
from scipy.spatial.distance import cosine
from scipy.stats import spearmanr
import numpy as np

# ------------------------------------------------------------
# 1. Load pretrained Word2Vec model (Google News)
# ------------------------------------------------------------
print("Loading pretrained 'word2vec-google-news-300' model (~1.6GB)...")
model_pretrained = api.load("word2vec-google-news-300")
print(f"Model loaded successfully! Vocabulary size: {len(model_pretrained)}")
print(f"Vector dimensions: {model_pretrained.vector_size}")

# ------------------------------------------------------------
# 2. Evaluate model on WordSim-353 (ensuring 353 pairs)
# ------------------------------------------------------------
def evaluate_model_on_wordsim(model, dataset):
    """
    Evaluate a pretrained Word2Vec model on WordSim-353.
    Computes cosine similarity for *all* 353 pairs.
    If a word is OOV (not in vocabulary), assigns similarity = 0.
    Returns Spearman correlation (ρ) and p-value.
    """
    predicted_sims, human_scores = [], []

    for _, row in dataset.iterrows():
        w1, w2 = row['Word 1'].lower(), row['Word 2'].lower()

        if w1 in model.key_to_index and w2 in model.key_to_index:
            sim = 1 - cosine(model[w1], model[w2])
        else:
            sim = 0.0  # handle out-of-vocabulary cases

        predicted_sims.append(sim)
        human_scores.append(row['Human (mean)'])

    # Compute Spearman correlation across all 353 pairs
    rho, p_value = spearmanr(predicted_sims, human_scores)
    return rho, p_value, len(predicted_sims)

# ------------------------------------------------------------
# 3. Run evaluation
# ------------------------------------------------------------
rho_pretrained, p_pretrained, n_pairs = evaluate_model_on_wordsim(model_pretrained, ws353)

print("\n" + "=" * 60)
print("STEP 5: PRETRAINED WORD2VEC EVALUATION")
print("=" * 60)
print(f"Spearman correlation (Google News pretrained): ρ = {rho_pretrained:.4f}")
print(f"P-value: {p_pretrained:.4e}")
print(f"Valid pairs: {n_pairs}/353")  # should now always show 353/353
print("=" * 60)


Loading pretrained 'word2vec-google-news-300' model (~1.6GB)...
[==================================================] 100.0% 1662.8/1662.8MB downloaded
Model loaded successfully! Vocabulary size: 3000000
Vector dimensions: 300

STEP 5: PRETRAINED WORD2VEC EVALUATION
Spearman correlation (Google News pretrained): ρ = 0.6833
P-value: 6.9317e-50
Valid pairs: 353/353


| Model                           | Dataset     | Spearman ρ | p-value          | Valid pairs   |
| ------------------------------- | ----------- | ---------- | ---------------- | ------------- |
| Google News pretrained Word2Vec | WordSim-353 | **0.6941** | **1.29 × 10⁻⁵¹** | **353 / 353** |


ρ = 0.6941 (p < 0.001)

The pretrained Google News Word2Vec model achieved a Spearman correlation coefficient of ρ = 0.6941 (p ≈ 1.29 × 10⁻⁵¹) on the WordSim-353 dataset.
This represents a strong, statistically significant positive correlation between model-predicted and human-rated similarity scores.

The result confirms that large-scale pretrained embeddings, trained on around 100 billion tokens, capture semantic relationships (e.g., plane–car, planet–sun) much more effectively than smaller, task-specific Word2Vec models trained in earlier steps.

**Part II: Hyperparameter tuning and Discussion of Results** **bold text**

In [14]:
# ---------------------------------------------------------
# Step 6: Hyperparameter tuning experiments
# ---------------------------------------------------------

# References:
# Mikolov, T., Chen, K., Corrado, G., & Dean, J. (2013).
#   Efficient Estimation of Word Representations in Vector Space.
#   arXiv preprint arXiv:1301.3781.  (Original Word2Vec paper)

# Merity, S., Xiong, C., Bradbury, J., & Socher, R. (2016).
#   Pointer Sentinel Mixture Models. arXiv:1609.07843.
#   (Example table structure used for reporting results)

# Levy, O., Goldberg, Y., & Dagan, I. (2015).
#   Improving Distributional Similarity with Lessons Learned from Word Embeddings.
#   Transactions of the Association for Computational Linguistics, 3, 211–225.
#   (Discussion of parameter effects like vector size, window, and corpus)

# Gensim Documentation:
#   https://radimrehurek.com/gensim/models/word2vec.html
import time
from gensim.models import Word2Vec
import pandas as pd
import numpy as np
from scipy.spatial.distance import cosine
from scipy.stats import spearmanr

# Computes Spearman correlation both for valid-only pairs and all 353 pairs

def evaluate_model_on_wordsim_dual(model, dataset):
    if hasattr(model, "wv"):
        key_index = model.wv.key_to_index
        get_vec = model.wv.__getitem__
    else:
        key_index = model.key_to_index
        get_vec = model.__getitem__

    predicted_valid, human_valid = [], []
    predicted_full, human_full = [], []

    for _, row in dataset.iterrows():
        w1, w2 = row["Word 1"].lower(), row["Word 2"].lower()
        human_score = row["Human (mean)"]

        if w1 in key_index and w2 in key_index:
            sim = 1 - cosine(get_vec(w1), get_vec(w2))
            predicted_valid.append(sim)
            human_valid.append(human_score)
            predicted_full.append(sim)
        else:
            # assign 0 for full version when missing
            predicted_full.append(0)

        human_full.append(human_score)

    # If no valid pairs, set NaN
    if len(predicted_valid) == 0:
        rho_valid, p_valid = np.nan, np.nan
    else:
        rho_valid, p_valid = spearmanr(predicted_valid, human_valid)

    rho_full, p_full = spearmanr(predicted_full, human_full)
    return {
        "rho_valid": rho_valid, "p_valid": p_valid, "n_valid": len(predicted_valid),
        "rho_full": rho_full, "p_full": p_full, "n_full": len(predicted_full)
    }

# Define experiments: you must run at least 2 extra experiments per dataset.
experiments = [
    # small corpus experiments (base + 2 variants)
    {"name": "small_base",     "data": train_small, "params": dict(vector_size=50,  window=5,  min_count=5,  epochs=5)},
    {"name": "small_vec100",   "data": train_small, "params": dict(vector_size=100, window=5,  min_count=5,  epochs=5)},
    {"name": "small_window10", "data": train_small, "params": dict(vector_size=50,  window=10, min_count=5,  epochs=5)},
    # large corpus experiments (base + 2 variants)
    {"name": "large_base",     "data": train_large, "params": dict(vector_size=50,  window=5,  min_count=5,  epochs=5)},
    {"name": "large_vec100",   "data": train_large, "params": dict(vector_size=100, window=5,  min_count=5,  epochs=5)},
    {"name": "large_epoch10",  "data": train_large, "params": dict(vector_size=50,  window=5,  min_count=5,  epochs=10)},
]

results = []

for exp in experiments:
    name = exp["name"]
    data = exp["data"]
    p = exp["params"]
    print(f"\n=== Training {name} | params: {p} ===")
    t0 = time.time()

    # Create model instance (don't pass sentences to constructor if you want to control build_vocab/train)
    model = Word2Vec(vector_size=p["vector_size"], window=p["window"], min_count=p["min_count"], workers=4)

    # build vocab and train explicitly (ensures epochs are used)
    model.build_vocab(data)
    model.train(data, total_examples=model.corpus_count, epochs=p["epochs"])

    t1 = time.time()
    train_time = t1 - t0

    # Evaluate (dual)
    eval_res = evaluate_model_on_wordsim_dual(model, ws353)

    # record stats
    row = {
        "Model": name,
        "corpus": "small" if "small" in name else "large",
        "vector_size": p["vector_size"],
        "window": p["window"],
        "min_count": p["min_count"],
        "epochs": p["epochs"],
        "vocab_size": len(model.wv),
        "train_time_s": round(train_time, 1),
        # valid-only metrics
        "rho_valid": np.round(eval_res["rho_valid"], 4) if not np.isnan(eval_res["rho_valid"]) else eval_res["rho_valid"],
        "p_valid": eval_res["p_valid"],
        "n_valid": eval_res["n_valid"],
        # full (353) metrics
        "rho_full": np.round(eval_res["rho_full"], 4),
        "p_full": eval_res["p_full"],
        "n_full": eval_res["n_full"]
    }
    results.append(row)
    print(f"Done. vocab_size={row['vocab_size']}, rho_valid={row['rho_valid']}, rho_full={row['rho_full']}")

# Convert to dataframe and show
df_results = pd.DataFrame(results)
# order columns nicely
cols = ["Model","corpus","vocab_size","vector_size","window","min_count","epochs","train_time_s",
        "n_valid","rho_valid","p_valid","n_full","rho_full","p_full"]
df_results = df_results[cols]
print("\n--- Experiment Results ---")
display(df_results)



=== Training small_base | params: {'vector_size': 50, 'window': 5, 'min_count': 5, 'epochs': 5} ===
Done. vocab_size=12291, rho_valid=0.1318, rho_full=-0.005

=== Training small_vec100 | params: {'vector_size': 100, 'window': 5, 'min_count': 5, 'epochs': 5} ===
Done. vocab_size=12291, rho_valid=0.1426, rho_full=0.0017

=== Training small_window10 | params: {'vector_size': 50, 'window': 10, 'min_count': 5, 'epochs': 5} ===
Done. vocab_size=12291, rho_valid=0.1426, rho_full=0.0

=== Training large_base | params: {'vector_size': 50, 'window': 5, 'min_count': 5, 'epochs': 5} ===
Done. vocab_size=148659, rho_valid=0.6398, rho_full=0.6161

=== Training large_vec100 | params: {'vector_size': 100, 'window': 5, 'min_count': 5, 'epochs': 5} ===
Done. vocab_size=148659, rho_valid=0.669, rho_full=0.6441

=== Training large_epoch10 | params: {'vector_size': 50, 'window': 5, 'min_count': 5, 'epochs': 10} ===
Done. vocab_size=148659, rho_valid=0.6431, rho_full=0.6202

--- Experiment Results ---


,Model,corpus,vocab_size,vector_size,window,min_count,epochs,train_time_s,n_valid,rho_valid,p_valid,n_full,rho_full,p_full
0,small_base,small,12291,50,5,5,5,1.1,266,0.1318,3.158674e-02,353,-0.0050,9.250959e-01
1,small_vec100,small,12291,100,5,5,5,1.2,266,0.1426,1.999568e-02,353,0.0017,9.750760e-01
2,small_window10,small,12291,50,10,5,5,1.3,266,0.1426,1.997693e-02,353,0.0000,9.995340e-01
3,large_base,large,148659,50,5,5,5,108.7,344,0.6398,5.276913e-41,353,0.6161,2.848377e-38
4,large_vec100,large,148659,100,5,5,5,122.0,344,0.6690,5.547405e-46,353,0.6441,9.219497e-43
5,large_epoch10,large,148659,50,5,5,10,211.7,344,0.6431,1.523989e-41,353,0.6202,6.723199e-39


**Part III – Word Embeddings Exploration: Analogies. Step 7 – Analogy Task**

In [15]:
# ============================================================
# STEP 7: Word Embeddings Exploration – Analogies
# ============================================================
# References:
# Mikolov, T., Chen, K., Corrado, G., & Dean, J. (2013).
#   Efficient Estimation of Word Representations in Vector Space.
#   arXiv preprint arXiv:1301.3781. (Introduced the parallelogram model for analogies)
#
# Rumelhart, D. E., & Abrahamson, A. A. (1973).
#   A model for analogical reasoning. Cognitive Psychology, 5(1), 1–28.
#   (Early cognitive model inspiring vector analogies)
#
# Bolukbasi, T., Chang, K.-W., Zou, J. Y., Saligrama, V., & Kalai, A. T. (2016).
#   Man is to Computer Programmer as Woman is to Homemaker? Debiasing Word Embeddings.
#   Advances in Neural Information Processing Systems (NeurIPS).
#   (Discusses gender bias in pretrained word embeddings)
#
# Gensim Documentation:
#   https://radimrehurek.com/gensim/models/keyedvectors.html#gensim.models.keyedvectors.KeyedVectors.most_similar
# ============================================================

import gensim.downloader as api

# ------------------------------------------------------------
# 1. Load Pretrained Google News Word2Vec Model
# ------------------------------------------------------------
print("Loading pretrained 'word2vec-google-news-300' model (~1.6GB)...")
model = api.load("word2vec-google-news-300")
print(" Model loaded successfully! Vocabulary size:", len(model))

# ------------------------------------------------------------
# 2. Helper Function for Analogies
# ------------------------------------------------------------
def show_analogies(model, positive_words, negative_words):
    """
    Implements the parallelogram model for analogies:
        'a' is to 'b' as 'c' is to ?
    Formula: vec(b) - vec(a) + vec(c)
    Returns top-5 most similar words.
    """
    try:
        results = model.most_similar(positive=positive_words, negative=negative_words, topn=5)
        print(f"\nAnalogy: {negative_words[0]} is to {positive_words[0]} as {positive_words[1]} is to ___ ?")
        print("Top-5 predictions:")
        for word, score in results:
            print(f"  {word:>15s}  (similarity = {score:.4f})")
        return results
    except KeyError as e:
        print(f"⚠️ Missing word in vocabulary: {e}")
        return None

# ------------------------------------------------------------
# 3. Analogy Tasks (Part g) – Semantic Relations
# ------------------------------------------------------------
print("\n=== Step 7 (g): Analogy Exploration – Semantic Analogies ===")

analogies_g = [
    (["woman", "king"], ["man"]),            # man:woman :: king:?
    (["Greece", "Rome"], ["Athens"]),        # Athens:Greece :: Rome:?
    (["read", "playing"], ["reading"]),      # reading:read :: playing:?
    (["souvlaki", "Italy"], ["Greece"]),     # Greece:souvlaki :: Italy:?
    (["propeller", "car"], ["airplane"])     # airplane:propeller :: car:?
]

results_g = []
for pos, neg in analogies_g:
    res = show_analogies(model, pos, neg)
    results_g.append((pos, neg, res))

# ------------------------------------------------------------
# 4. Analogy Tasks (Part h) – Gender Bias Exploration
# ------------------------------------------------------------
print("\n=== Step 7 (h): Analogy Exploration – Gender Bias Tests ===")

analogies_h = [
    (["woman", "computer_programmer"], ["man"]),
    (["woman", "superstar"], ["man"]),
    (["woman", "guitarist"], ["man"]),
    (["woman", "boss"], ["man"])
]

results_h = []
for pos, neg in analogies_h:
    res = show_analogies(model, pos, neg)
    results_h.append((pos, neg, res))


Loading pretrained 'word2vec-google-news-300' model (~1.6GB)...
 Model loaded successfully! Vocabulary size: 3000000

=== Step 7 (g): Analogy Exploration – Semantic Analogies ===

Analogy: man is to woman as king is to ___ ?
Top-5 predictions:
            queen  (similarity = 0.7118)
          monarch  (similarity = 0.6190)
         princess  (similarity = 0.5902)
     crown_prince  (similarity = 0.5499)
           prince  (similarity = 0.5377)

Analogy: Athens is to Greece as Rome is to ___ ?
Top-5 predictions:
            Italy  (similarity = 0.6826)
           Sicily  (similarity = 0.5808)
         Portugal  (similarity = 0.5467)
          Italian  (similarity = 0.5194)
             ANSA  (similarity = 0.5114)

Analogy: reading is to read as playing is to ___ ?
Top-5 predictions:
           played  (similarity = 0.7010)
             play  (similarity = 0.6676)
          Playing  (similarity = 0.5651)
           playin  (similarity = 0.5131)
           toplay  (similarity = 0.4870)



The pretrained Google News Word2Vec model exhibits gender bias, confirming findings by Bolukbasi et al. (2016). These biases likely reflect stereotypes in the massive web-text training data (e.g., associating women with homemaking or support roles).

While analogies like king → queen show legitimate semantic structure, biased analogies highlight the need for debiasing techniques before applying pretrained embeddings in sensitive contexts.